File: test-src/16_listExtensions.kt
import java.util.*

fun Float.timesTen(): Float {
  return this * 10
}

fun List<Float>.avg(): Float {
  if (this.size == 0) {
    return 0f
  }

  var sum = 0.0f
  for (item in this) {
    // SWIFT: sum += item as! Double
    sum += item
  }
  return sum / this.size
}

class Person(val name: String, val age: Int) {
}

fun List<Person>.countAdults(): Int {
  var adultCounter = 0
  for (person in this) {
    if (person.age >= 18) {
      adultCounter++
    }
  }
  return adultCounter
}

fun main(args: Array<String>) {
  val list = ArrayList<Float>()
  list.add(1.5f.timesTen())
  list.add(1f.timesTen())
  list.add(11.858502f)
  list.add(3.1415f)
  val avg1 = list.avg()
  println("avg1 = $avg1 (should be 10.0)")
    
  val list2 = LinkedList<Float>()
  list2.add(15.0f)
  list2.add(5.0f)
  println("avg2 = ${list2.avg()} (should be 10.0)")

  val people = ArrayList<Person>()
  people.add(Person(name = "Steve", age = 14))
  people.add(Person(name = "Bob", age = 16))
  people.add(Person(name = "John", age = 18))
  people.add(Person(name = "Lena", age = 20))
  people.add(Person(name = "Denise", age = 22))
  people.add(Person(name = "Alex", age = 24))
  println("${people.countAdults()} people may enter the club (should be 4)")
}

File: test-dest/16_listExtensions.swift

extension Double {
func timesTen() -> Double {
  return self * 10
}
}

extension Array where Element : Addable {
func avg() -> Double {
  if self.size == 0 {
    return 0
  }

  var sum = 0.0
  for item in self {
    sum += item as! Double
  }
  return sum / self.size
}
}

public class Person {
  let name: String
  let age: Int32

  init(name: String, age: Int32) {
    self.name = name
    self.age = age
  }
}

extension Array where Element : Person {
func countAdults() -> Int32 {
  var adultCounter = 0
  for person in self {
    if person.age >= 18 {
      adultCounter++
    }
  }
  return adultCounter
}
}

func main(args: [String]) {
  var list = [Double]()
  list.add(1.5.timesTen())
  list.add(1.timesTen())
  list.add(11.858502)
  list.add(3.1415)
  let avg1 = list.avg()
  print("avg1: \(avg1) (should be 10.0)")

  var list2 = [Double]()
  list2.add(15.0)
  list2.add(5.0)
  print("avg2: \(list2.avg()) (should be 10.0)")

  var people = [Person]()
  people.add(Person(name: "Steve", age: 14))
  people.add(Person(name: "Bob", age: 16))
  people.add(Person(name: "John", age: 18))
  people.add(Person(name: "Lena", age: 20))
  people.add(Person(name: "Denise", age: 22))
  people.add(Person(name: "Alex", age: 24))
  print("\(people.countAdults()) people may enter the club (should be 4)")
}

main([])
